### PR TITLE
Comments on Translclass

### PR DESCRIPTION
**Pull request by lthls**
This PR ports a good chunk of the comments @Ekdohibs and I wrote on the compilation of objects (the rest can be seen at https://github.com/lthls/ocaml/tree/objects_documentation).
I will try to add some of the comments on `CamlinternalOO` too, although I think this PR can be reviewed (and merged) independently.

### CODE REVIEW

**Comment by OlivierNicole** [comment]
I love having more documentation comments, thanks!

I’m putting this on my review list, but it might take me a while to get to it, so others should not hesitate to review (I’m also not the most well-versed on objects).

File: lambda/translclass.ml [hunk 1]
```diff
@@ -13,6 +13,176 @@
...
+   Objects are represented in memory using two layers:
+   - The outer layer is a block with tag [Obj.object_tag].
+     It has a first field pointing to the inner layer (the methods),
+     a second field acting as a unique identifier to allow
+     polymorphic comparison, and the rest of the block contains
+     the values of the instance variables, class parameters, and
+     other values that can vary between two objects of the same class.
+
+   - The inner layer is a regular block (with tag zero). It contains
+     all values that are shared between all objects of the same classes,
```
Comment (OlivierNicole) [inline lambda/translclass.ml]:
```suggestion
     all values that are shared between all objects of the same class,
```

File: lambda/translclass.ml [hunk 2]
```diff
@@ -200,15 +439,20 @@ let rec build_object_init ~scopes cl_table obj params inh_init obj_init cl =
...
   | Tcl_let (rec_flag, defs, vals, cl) ->
+      (* See comment on the [Tcl_fun] case for the meaning of [vals] *)
       let (inh_init, obj_init) =
         build_object_init ~scopes cl_table obj (vals @ params)
           inh_init obj_init cl
       in
       (inh_init, Translcore.transl_let ~scopes rec_flag defs obj_init)
   | Tcl_open (_, cl)
+    (* Class local opens are restricted to paths only,
+       so no code is generated *)
```
Comment (OlivierNicole) [inline lambda/translclass.ml]:
```suggestion
    (* Class local opens are restricted to paths only, so no code is generated
     *)
```

**Review by OlivierNicole — APPROVED** [review]
This looks good to me, and I’m glad to have more documentation of the code.

Disclaimer: I am by no means an object specialist. What I checked is:
- that the high-level description of the algorithm makes sense and corresponds to what I read from other sources, notably <https://ambassadortothecomputers.blogspot.com/2010/03/inside-ocaml-objects.html>;
- that the added comments in the implementation seem to make sense in the context of the code around them.

I did _not_ check that every comment in the implementation describes exactly what happens, because it would be a very long effort for me. IMO unless someone volunteers to do that, the risks/benefits ratio is largely in favor of merging this.

**Comment by gasche** [comment]
@OlivierNicole I agree with your risk/benefit analysis, and I would encourage you to merge based on this assessment.

**Comment by lthls** [comment]
Thanks ! I've taken your suggestions into account, and I plan to click merge on Monday if the CI doesn't find anything to complain about.

**Comment by lthls** [comment]
(I'm fine with merging immediately too)

**Comment by OlivierNicole** [comment]
I went ahead and merged. Less cognitive load for Monday!

**Pull request merged**